Issue from Android Studio

Please file this bug report at https://github.com/flutter/flutter-intellij/issues/new.

---

## What happened

I pressed on "Open Android Emulator Pixel 5" up top the taskbar of the available devices

## Version information

Android Studio `Dolphin | 2021.3.1` • Flutter plugin `io.flutter 70.2.3` • Dart plugin `213.7433`

[√] Flutter (Channel stable, 3.3.3, on Microsoft Windows [Versione 10.0.19044.2130], locale it-IT)
[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc1)              
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Build Tools 2019 16.11.9)
[√] Android Studio (version 2021.2)
[√] Android Studio (version 2021.3)
[√] VS Code (version 1.72.2)
[√] Connected device (4 available)
[√] HTTP Host Availability

• No issues found!

## Exception

null

```
java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.GeneratedMethodAccessor108.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.intellij.execution.process.ProcessHandler$2.invoke(ProcessHandler.java:247)
	at com.sun.proxy.$Proxy38.onTextAvailable(Unknown Source)
	at com.intellij.execution.process.ProcessHandler.notifyTextAvailable(ProcessHandler.java:221)
	at com.intellij.execution.process.ColoredProcessHandler.coloredTextAvailable(ColoredProcessHandler.java:68)
	at io.flutter.utils.MostlySilentColoredProcessHandler.coloredTextAvailable(MostlySilentColoredProcessHandler.java:77)
	at com.intellij.execution.process.AnsiEscapeDecoder.processTextChunk(AnsiEscapeDecoder.java:100)
	at com.intellij.execution.process.AnsiEscapeDecoder.processTextWithoutAnsi(AnsiEscapeDecoder.java:81)
	at com.intellij.execution.process.AnsiEscapeDecoder.escapeText(AnsiEscapeDecoder.java:48)
	at com.intellij.execution.process.ColoredProcessHandler.notifyTextAvailable(ColoredProcessHandler.java:57)
	at com.intellij.execution.process.BaseOSProcessHandler.startNotify(BaseOSProcessHandler.java:75)
	at io.flutter.android.AndroidEmulator.startEmulator(AndroidEmulator.java:93)
	at io.flutter.actions.OpenEmulatorAction.actionPerformed(OpenEmulatorAction.java:44)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.lambda$performActionDumbAwareWithCallbacks$4(ActionUtil.java:239)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.java:260)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAwareWithCallbacks(ActionUtil.java:239)
	at com.intellij.ui.popup.ActionPopupStep.performAction(ActionPopupStep.java:253)
	at com.intellij.ui.popup.ActionPopupStep.performAction(ActionPopupStep.java:243)
	at com.intellij.ui.popup.ActionPopupStep.lambda$onChosen$2(ActionPopupStep.java:229)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:94)
	at com.intellij.ui.popup.AbstractPopup.lambda$dispose$18(AbstractPopup.java:1508)
	at com.intellij.util.ui.EdtInvocationManager.invokeLaterIfNeeded(EdtInvocationManager.java:101)
	at com.intellij.ide.IdeEventQueue.ifFocusEventsInTheQueue(IdeEventQueue.java:186)
	at com.intellij.ide.IdeEventQueue.executeWhenAllFocusEventsLeftTheQueue(IdeEventQueue.java:140)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.doWhenFocusSettlesDown(FocusManagerImpl.java:175)
	at com.intellij.openapi.wm.impl.IdeFocusManagerImpl.doWhenFocusSettlesDown(IdeFocusManagerImpl.java:36)
	at com.intellij.ui.popup.AbstractPopup.dispose(AbstractPopup.java:1505)
	at com.intellij.ui.popup.WizardPopup.dispose(WizardPopup.java:164)
	at com.intellij.ui.popup.list.ListPopupImpl.dispose(ListPopupImpl.java:326)
	at com.intellij.ui.popup.PopupFactoryImpl$ActionGroupPopup.dispose(PopupFactoryImpl.java:271)
	at com.intellij.openapi.util.ObjectTree.runWithTrace(ObjectTree.java:139)
	at com.intellij.openapi.util.ObjectTree.executeAll(ObjectTree.java:169)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:217)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:205)
	at com.intellij.ui.popup.WizardPopup.disposeAllParents(WizardPopup.java:268)
	at com.intellij.ui.popup.list.ListPopupImpl.handleNextStep(ListPopupImpl.java:433)
	at com.intellij.ui.popup.list.ListPopupImpl._handleSelect(ListPopupImpl.java:405)
	at com.intellij.ui.popup.list.ListPopupImpl.handleSelect(ListPopupImpl.java:361)
	at com.intellij.ui.popup.PopupFactoryImpl$ActionGroupPopup.handleSelect(PopupFactoryImpl.java:288)
	at com.intellij.ui.popup.list.ListPopupImpl$MyMouseListener.mouseReleased(ListPopupImpl.java:618)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:298)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at com.intellij.ui.popup.list.ListPopupImpl$MyList.processMouseEvent(ListPopupImpl.java:695)
	at java.desktop/java.awt.Component.processEvent(Component.java:6419)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5029)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2790)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:820)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:757)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:498)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: com.intellij.serviceContainer.AlreadyDisposedException: Cannot create com.intellij.openapi.wm.ToolWindowManager because container is already disposed (container=Project(name=dart-ristocloud, containerState=DISPOSE_COMPLETED, componentStore=D:\Ristocloud\dart-ristocloud) (disposed))
	at com.intellij.serviceContainer.ContainerUtilKt.throwAlreadyDisposedError(containerUtil.kt:34)
	at com.intellij.serviceContainer.ComponentManagerImpl.doGetService(ComponentManagerImpl.kt:617)
	at com.intellij.serviceContainer.ComponentManagerImpl.getService(ComponentManagerImpl.kt:569)
	at com.intellij.openapi.client.ClientAwareComponentManager.getFromSelfOrCurrentSession(ClientAwareComponentManager.kt:37)
	at com.intellij.openapi.client.ClientAwareComponentManager.getService(ClientAwareComponentManager.kt:22)
	at com.intellij.openapi.wm.ToolWindowManager$Companion.getInstance(ToolWindowManager.kt:26)
	at com.intellij.openapi.wm.ToolWindowManager.getInstance(ToolWindowManager.kt)
	at io.flutter.android.AndroidEmulator.openEmulatorToolWindow(AndroidEmulator.java:105)
	at io.flutter.android.AndroidEmulator$1.onTextAvailable(AndroidEmulator.java:80)
	... 82 more
```

--- a/flutter-idea/src/io/flutter/actions/OpenEmulatorAction.java
+++ b/flutter-idea/src/io/flutter/actions/OpenEmulatorAction.java
@@ -12,6 +12,7 @@ import io.flutter.android.AndroidEmulator;
 import io.flutter.sdk.AndroidEmulatorManager;
 import org.jetbrains.annotations.NotNull;
 
+import java.util.ArrayList;
 import java.util.List;
 
 import static java.util.stream.Collectors.toList;
@@ -25,6 +26,9 @@ public class OpenEmulatorAction extends AnAction {
    * to the list of emulators.
    */
   public static List<OpenEmulatorAction> getEmulatorActions(Project project) {
+    if (project == null || project.isDisposed()) {
+      return new ArrayList<>();
+    }
     final AndroidEmulatorManager emulatorManager = AndroidEmulatorManager.getInstance(project);
 
     final List<AndroidEmulator> emulators = emulatorManager.getCachedEmulators();

--- a/flutter-idea/src/io/flutter/android/AndroidEmulator.java
+++ b/flutter-idea/src/io/flutter/android/AndroidEmulator.java
@@ -23,17 +23,14 @@ import io.flutter.FlutterMessages;
 import io.flutter.utils.MostlySilentColoredProcessHandler;
 import org.jetbrains.annotations.NotNull;
 
-import java.util.Objects;
-import java.util.Timer;
-import java.util.TimerTask;
-
 public class AndroidEmulator {
   private static final Logger LOG = Logger.getInstance(AndroidEmulator.class);
 
-  final AndroidSdk androidSdk;
-  final String id;
+  @NotNull final AndroidSdk androidSdk;
+  @NotNull final String id;
+  ProcessAdapter listener;
 
-  AndroidEmulator(AndroidSdk androidSdk, String id) {
+  AndroidEmulator(@NotNull AndroidSdk androidSdk, @NotNull String id) {
     this.androidSdk = androidSdk;
     this.id = id;
   }
@@ -43,6 +40,9 @@ public class AndroidEmulator {
   }
 
   public void startEmulator() {
+    if (androidSdk.project.isDisposed()) {
+      return;
+    }
     final VirtualFile emulator = androidSdk.getEmulatorToolExecutable();
     if (emulator == null) {
       FlutterMessages.showError(
@@ -71,9 +71,12 @@ public class AndroidEmulator {
     try {
       final StringBuilder stdout = new StringBuilder();
       final ColoredProcessHandler process = new MostlySilentColoredProcessHandler(cmd);
-      process.addProcessListener(new ProcessAdapter() {
+      listener = new ProcessAdapter() {
         @Override
         public void onTextAvailable(@NotNull ProcessEvent event, @NotNull Key outputType) {
+          if (androidSdk.project.isDisposed()) {
+            return;
+          }
           if (outputType == ProcessOutputTypes.STDERR || outputType == ProcessOutputTypes.STDOUT) {
             stdout.append(event.getText());
           }
@@ -81,6 +84,7 @@ public class AndroidEmulator {
         }
 
         public void processTerminated(@NotNull ProcessEvent event) {
+          process.removeProcessListener(listener);
           final int exitCode = event.getExitCode();
           if (exitCode != 0) {
             final String message = stdout.length() == 0
@@ -89,7 +93,8 @@ public class AndroidEmulator {
             FlutterMessages.showError("Error Opening Emulator", message, androidSdk.project);
           }
         }
-      });
+      };
+      process.addProcessListener(listener);
       process.startNotify();
     }
     catch (ExecutionException | RuntimeException e) {
@@ -101,7 +106,7 @@ public class AndroidEmulator {
     if (!shouldLaunchEmulatorInToolWindow) {
       return;
     }
-    if (androidSdk == null || androidSdk.project == null || androidSdk.project.isDisposed()) {
+    if (androidSdk.project.isDisposed()) {
       return;
     }
     final ToolWindowManager wm = ToolWindowManager.getInstance(androidSdk.project);

